tests/rt_current_thread.rs

```diff
@@ -164,3 +164,29 @@ fn idle_after_block_on() {
     rt.run().unwrap();
     assert!(ran.load(Ordering::SeqCst));
 }
+
+#[test]
+fn enter_exposed() {
+    let rt = current_thread::Runtime::new().unwrap();
+    rt.enter(|| {
+        let _handle = tokio_02::runtime::Handle::current();
+    });
+}
+
+#[test]
+fn enter_can_spawn_01_futures() {
+    use futures_01::future::IntoFuture;
+    let future_ran = Arc::new(AtomicBool::new(false));
+    let ran = future_ran.clone();
+    let rt = current_thread::Runtime::new().unwrap();
+    rt.enter(|| {
+        let future = tokio_01::spawn(futures_01::future::lazy(move || {
+            future_ran.store(true, Ordering::SeqCst);
+            Ok(())
+        }))
+        .into_future()
+        .compat();
+        futures_03::executor::block_on(future)
+    }).unwrap();
```

Comment:
I think the reason this test is failing is because we are `spawn`ing the future. This means it is now running _on the `tokio-compat` runtime_. The call to `tokio` 0.1's `spawn` function returns a `Result`. `Result` implements `IntoFuture` by returning a future that finishes _immediately_ with the `Ok` or `Err` value, so when we call `into_future`, we get back a completed `Ok(())` future. Then, we `block_on` that future, which also returns immediately. However, the _spawned future_, which is running on the `tokio-compat` runtime, is never run, since we never actually run that runtime's futures (such as with a call to `Runtime::block_on` or `Runtime::run`/`shutdown_on_idle`).

I think if the test was changed to something like this, it would pass:
```rust
    let future_ran = Arc::new(AtomicBool::new(false));
    let ran = future_ran.clone();
    let rt = current_thread::Runtime::new().unwrap();
    rt.enter(|| {
        let future = futures_01::future::lazy(move || {
            future_ran.store(true, Ordering::SeqCst);
            Ok(())
        });
        futures_03::executor::block_on(future)
    }).unwrap();
    assert!(ran.load(Ordering::SeqCst));
```
However, this wouldn't really test the behavior of `enter`, since the same code would work with the `tokio_compat` runtime totally removed. I think what we would really want to be testing is whether code inside of the `enter` can spawn futures on the entered runtime that _will_ be run when that runtime is driven. We might want to have a test like this:
```rust
    let future1_ran = Arc::new(AtomicBool::new(false));
    let ran1 = future1_ran.clone();
    let future2_ran = Arc::new(AtomicBool::new(false));
    let ran2 = future2_ran.clone();
    let rt = current_thread::Runtime::new().unwrap();
    rt.enter(|| {
        let future = futures_01::future::lazy(move || {
            tokio_01::spawn(futures_01::future::lazy(move || {
                future2_ran.store(true, Ordering::SeqCst);
                Ok(())
            }))?;    
            future1_ran.store(true, Ordering::SeqCst);
            Ok(())
        });
        futures_03::executor::block_on(future)
    }).unwrap();
    assert!(ran1.load(Ordering::SeqCst));

    // The runtime should run any futures spawned in the `enter` block.
    rt.run();
    assert!(ran2.load(Ordering::SeqCst));
```

I'm not convinced the `futures_03::executor::block_on` is strictly necessary for this test, so I think we can simplify this to just something like this:
```rust
    let future_ran = Arc::new(AtomicBool::new(false));
    let ran = future_ran.clone();
    let rt = current_thread::Runtime::new().unwrap();
    rt.enter(|| {
        tokio_01::spawn(futures_01::future::lazy(move || {
            future_ran.store(true, Ordering::SeqCst);
            Ok(())
        })).unwrap();
    });

    // The runtime should run any futures spawned in the `enter` block.
    rt.run();
    assert!(ran1.load(Ordering::SeqCst));
```
Does that make sense?

Reply:
Yeah, that makes sense. For whatever reason, I thought [`Spawn`](https://docs.rs/tokio/0.1.19/tokio/executor/struct.Spawn.html) was something that could be awaited. It can't.

Reply:
In `tokio` 0.2, [it is](https://docs.rs/tokio/0.2.9/tokio/task/index.html#spawning), but not in `tokio` 0.1 :)